src/Sentry/Internal/BackpressureMonitor.cs

```diff
@@ -0,0 +1,166 @@
+using Sentry.Extensibility;
+using Sentry.Infrastructure;
+
+namespace Sentry.Internal;
+
+/// <summary>
+/// <para>
+/// Monitors system health and calculates a DownsampleFactor that can be applied to events and transactions when the
+/// system is under load.
+/// </para>
+/// <para>
+/// The health checks used by the monitor are:
+/// </para>
+/// <list type="number">
+///     <item>if any events have been dropped due to queue being full in the last 2 seconds</item>
+///     <item>if any new rate limits have been applied since the last check</item>
+/// </list>
+/// This check is performed every 10 seconds. With each negative health check we halve tracesSampleRate up to 10 times, meaning the original tracesSampleRate is multiplied by 1, 1/2, 1/4, ... up to 1/1024 (~ 0.001%). Any positive health check resets to the original tracesSampleRate set in SentryOptions.
+/// </summary>
+/// <seealso href="https://develop.sentry.dev/sdk/telemetry/traces/backpressure/">Backpressure Management</seealso>
+internal class BackpressureMonitor : IDisposable
+{
+    internal const int MaxDownsamples = 10;
+    private const int CheckIntervalInSeconds = 10;
+    private const int RecentThresholdInSeconds = 2;
+
+    private readonly IDiagnosticLogger? _logger;
+    private readonly ISystemClock _clock;
+    private long _lastQueueOverflow = DateTimeOffset.MinValue.Ticks;
+    private long _lastRateLimitEvent = DateTimeOffset.MinValue.Ticks;
+    private volatile int _downsampleLevel = 0;
+
+    private static readonly long RecencyThresholdTicks = TimeSpan.FromSeconds(RecentThresholdInSeconds).Ticks;
+    private static readonly long CheckIntervalTicks = TimeSpan.FromSeconds(CheckIntervalInSeconds).Ticks;
+
+    private readonly CancellationTokenSource _cts = new();
+
+    private Task WorkerTask { get; }
+    internal int DownsampleLevel => _downsampleLevel;
+    internal long LastQueueOverflowTicks => Interlocked.Read(ref _lastQueueOverflow);
+    internal long LastRateLimitEventTicks => Interlocked.Read(ref _lastRateLimitEvent);
+
+    public BackpressureMonitor(IDiagnosticLogger? logger, ISystemClock? clock = null, bool startImmediately = true)
```

Comment:
```diff
-bool startImmediately = true
+bool enablePeriodicHealthCheck = true
```

To me `startImmediately` is a bit misleading, as there is no way to start the Periodic-Health-Check after construction of the Monitor.

I suggest to rename `startImmediately` to something like
- `enablePeriodicHealthCheck`
- `startPeriodicHealthCheck`
- `withPeriodicHealthCheck`

or similar.